tespy/components/customs.py

```diff
@@ -0,0 +1,818 @@
+        mat_deriv : ndarray
+            Matrix of partial derivatives.
+        """
+
+        ######################################################################
+        # derivatives for saturated liquid at hot side 1 outlet equation
+        if self.subcooling.val is False:
+            o1 = self.outl[0].to_flow()
+            self.mat_deriv[k, 3, 1] = -dh_mix_dpQ(o1, 0)
+            self.mat_deriv[k, 3, 2] = 1
+            k += 1
+
+        ######################################################################
+        # derivatives for saturated gas at cold side outlet 3 equation
+        if self.overheating.val is False:
+            o3 = self.outl[2].to_flow()
+            self.mat_deriv[k, 5, 1] = -dh_mix_dpQ(o3, 1)
+            self.mat_deriv[k, 5, 2] = 1
+            k += 1
+
+    def fluid_func(self):
+        r"""
+        Calculates the vector of residual values for component's fluid balance
+        equations.
+
+        Returns
+        -------
+        vec_res : list
+            Vector of residual values for component's fluid balance.
+
+            .. math::
+
+                0 = fluid_{i,in_{j}} - fluid_{i,out_{j}} \;
+                \forall i \in \mathrm{fluid}, \; \forall j \in inlets/outlets
+        """
+        vec_res = []
+
+        for i in range(self.num_i):
+            for fluid, x in self.inl[i].fluid.val.items():
+                vec_res += [x - self.outl[i].fluid.val[fluid]]
+        return vec_res
+
+    def mass_flow_func(self):
+        r"""
+        Calculates the residual value for component's mass flow balance
+        equation.
+
+        Returns
+        -------
+        vec_res : list
+            Vector with residual value for component's mass flow balance.
+
+            .. math::
+
+                0 = \dot{m}_{in,i} - \dot{m}_{out,i} \;
+                \forall i \in inlets/outlets
+        """
+        vec_res = []
+        for i in range(self.num_i):
+            vec_res += [self.inl[i].m.val_SI - self.outl[i].m.val_SI]
+        return vec_res
+
+    def fluid_deriv(self):
+        r"""
+        Calculates the partial derivatives for all fluid balance equations.
+
+        Returns
+        -------
+        deriv : list
+            Matrix with partial derivatives for the fluid equations.
+        """
+        deriv = np.zeros((self.num_nw_fluids * 3,
+                          6 + self.num_vars,
+                          self.num_nw_vars))
+        # hot side 1
+        i = 0
+        for fluid in self.nw_fluids:
+            deriv[i, 0, i + 3] = 1
+            deriv[i, 3, i + 3] = -1
+            i += 1
+        # hot side 2
+        j = 0
+        for fluid in self.nw_fluids:
+            deriv[i + j, 1, j + 3] = 1
+            deriv[i + j, 4, j + 3] = -1
+            j += 1
+        # cold side
+        k = 0
+        for fluid in self.nw_fluids:
+            deriv[i + j + k, 2, k + 3] = 1
+            deriv[i + j + k, 5, k + 3] = -1
+            k += 1
+        return deriv.tolist()
+
```

Comment:
This is quite long. The is True for the implementation of the heat_exchanger class as well. We could change the base functionality of the fluid_deriv in the component class in order to work with all components that do not change fluid properties. The implementation of the fluid_deriv in the subsystem_interface class looks good for this purpose.
```
    def fluid_deriv(self):
        r"""
        Calculate the partial derivatives for all fluid balance equations.
        Returns
        -------
        deriv : list
            Matrix with partial derivatives for the fluid equations.
        """
        deriv = np.zeros((self.num_nw_fluids * self.num_i,
                          2 * self.num_i,
                          self.num_nw_vars))
        for i in range(self.num_i):
            for j in range(self.num_nw_fluids):
                deriv[i * self.num_nw_fluids + j, i, j + 3] = 1
                deriv[i * self.num_nw_fluids + j, self.num_i + i, j + 3] = -1
        return deriv.tolist()
```

Reply:
Yes, thanks for that.